components/eam/src/physics/cam/cldera_dynamic_tracers.F90

```diff
@@ -0,0 +1,371 @@
+
+    !---------------------------Local workspace-----------------------------
+    integer :: m
+    !-----------------------------------------------------------------------
+
+    cldera_dynamic_tracers_implements_cnst = .false.
+
+    if (.not. cldera_dynamic_tracers_flag) return
+
+    do m = 1, ncnst
+       if (name == c_names(m)) then
+          cldera_dynamic_tracers_implements_cnst = .true.
+          return
+       end if
+    end do
+
+  end function cldera_dynamic_tracers_implements_cnst
+
+!===============================================================================
+
+  subroutine cldera_dynamic_tracers_pv_idx(idx)
+    !----------------------------------------------------------------------- 
+    ! 
+    ! Purpose: getter fucntion returning global constituent index of PV tracer
+    ! 
+    !-----------------------------------------------------------------------
+    integer, intent(out) :: idx
+
+    idx = 0
+    if (.not. cldera_dynamic_tracers_flag) return
+
+    idx = ixpv
+
+
+  end subroutine cldera_dynamic_tracers_pv_idx
+  
+!===============================================================================
+
+  subroutine cldera_dynamic_tracers_pt_idx(idx)
+    !----------------------------------------------------------------------- 
+    ! 
+    ! Purpose: getter fucntion returning global constituent index of PT tracer
+    ! 
+    !-----------------------------------------------------------------------
+    integer, intent(out) :: idx
+
+    idx = 0
+    if (.not. cldera_dynamic_tracers_flag) return
+
+    idx = ixpt
+
+
+  end subroutine cldera_dynamic_tracers_pt_idx
+
+!===============================================================================
+
+  subroutine cldera_dynamic_tracers_timestep_tend(state, ptend, dt, ncol)
+
+    use physics_types, only: physics_state, physics_ptend, physics_ptend_init
+    use phys_grid,     only: get_rlat_all_p , get_lat_all_p
+    use physconst,     only: gravit, avogad
+    use cam_history,   only: outfld
+    use time_manager,  only: get_nstep
+    use ref_pres,      only: pref_mid_norm
+    use time_manager,  only: get_curr_time
+
+    ! Arguments
+    type(physics_state), intent(inout) :: state              ! state variables
+    type(physics_ptend), intent(out)   :: ptend              ! package tendencies
+    real(r8),            intent(in)    :: dt                 ! timestep
+    integer,             intent(in)    :: ncol
+
+    !----------------- Local workspace-------------------------------
+
+    integer :: i, k
+    integer :: lchnk             ! chunk identifier
+
+    real(r8) :: pv_out(ncol,pver)
+    if (.not. cldera_dynamic_tracers_flag) then
+      return
+    end if
+    lchnk = state%lchnk
+    do i=1,ncol
+      do k=1,pver
+        pv_out(i,k) = state%q(i,k,ixpv)
+      end do
+    end do
+    call outfld(pv_out_name, pv_out(:,:)-cldera_dynamic_tracers_pv_offset, ncol, lchnk)           
```

Comment:
Just flagging this as something to check if runs with debug symbols crash with segfaults. We sometimes have trouble with this pattern; pv_out has dimensions (pcols,pver), and pcol >= ncol. I think since you’ve passed ncol it might be okay but need to check on this.

Reply:
I'm not sure I understand; looks like pv_out is defined as `real(r8) :: pv_out(ncol,pver)`. I don't see where we're using pcol. 

Also, a naive question... if ncol is referencing a physics column, is pcol referencing a dynamics column? I assume so since you say pcol >= ncol.

Reply:
No, these are all physics columns. But due to the decomposition for mpi, some ranks might have slightly more columns than others. So pcols is the maximum number of columns any one rank will get, and ncol is the actual number of physics columns a given rank is operating on. These numbers may be slightly different depending on how the number of available ranks divides the number of elements. This can cause some issues if these aren't handled quite right in the code, and the Fortran standard lets you do some weird things that makes this worse. You're right about pv_out, I don't know why I read that as having dim pcols, so you're probably fine here.

Reply:
My reading is that the issues referenced here are resolved, is that correct? 
If the bound that I've called `ncol` here is usually called `pcol` on the physics side of things, I would be unopposed to changing it.

Reply:
No issues here, just making a note.